x-pack/legacy/plugins/fleet/public/components/agent_unenroll_provider.tsx

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Fragment, useRef, useState } from 'react';
+import { EuiConfirmModal, EuiOverlayMask } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { useLibs } from '../hooks/use_libs';
+
+interface Props {
+  children: (unenrollAgents: UnenrollAgents) => React.ReactElement;
+}
+
+export type UnenrollAgents = (
+  agents: string[] | string,
+  agentsCount: number,
+  onSuccess?: OnSuccessCallback
+) => void;
+
+type OnSuccessCallback = (agentsUnenrolled: string[]) => void;
+
+export const AgentUnenrollProvider: React.FunctionComponent<Props> = ({ children }) => {
+  const libs = useLibs();
+  const [agents, setAgents] = useState<string[] | string>([]);
```

Comment:
happy to discuss that, but I have a strong preference to use only one state `{ agents, agentCount, isLoading ...}`:
* With typescript you can define all the possible variant of your state and avoid illegal state for example `{isLoading: true, isModalOpen: false}`
* You avoid too many `setState` and component re render 

Also personal preference I like to move the state logic in a hook outside the component, I think it make thing more readable

Reply:
🤔 I norm write my code like @jen-huang did in this file, but you make REALLY good points here. I am a fan. Moving state logic out would make it much easier to test the logic in unit tests so I am especially fond of that.

Reply:
I think the typescript variants and unit testing are great benefits; I'm not worried about re-rendering since 1) React batches multiple `setState` calls and 2) re-rendering first happens in virtual DOM before being pushed up to real DOM

I'll try moving state logic out the next chance I have and see if I like it 😄